consensus/core/src/synchronizer.rs

```diff
@@ -649,6 +692,111 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
         (resp, blocks_guard, retries, peer, highest_rounds)
     }
 
+    fn start_fetch_own_last_block_task(&mut self) {
+        const FETCH_OWN_BLOCK_RETRY_DELAY: Duration = Duration::from_millis(1_000);
+
+        let context = self.context.clone();
+        let network_client = self.network_client.clone();
+        let block_verifier = self.block_verifier.clone();
+        let core_dispatcher = self.core_dispatcher.clone();
+
+        self.fetch_own_last_block_task
+            .spawn(monitored_future!(async move {
+                let _scope = monitored_scope("FetchOwnLastBlockTask");
+
+                // Ask all the other peers about our last block
+                let mut results = FuturesUnordered::new();
+
+                let fetch_own_block = |authority_index: AuthorityIndex, fetch_own_block_delay: Duration| {
+                    let network_client_cloned = network_client.clone();
+                    let context_cloned = context.clone();
+                    async move {
+                        sleep(fetch_own_block_delay).await;
+                        let r = network_client_cloned.fetch_latest_blocks(authority_index, vec![context_cloned.own_index], FETCH_REQUEST_TIMEOUT).await;
+                        (r, authority_index)
+                    }
+                };
+
+                for (authority_index, _authority) in context.committee.authorities() {
+                    if authority_index != context.own_index {
+                        results.push(fetch_own_block(authority_index, Duration::from_millis(0)));
+                    }
+                }
+
+                // Gather the results but wait to timeout as well
+                let timer = sleep_until(Instant::now() + context.parameters.sync_last_proposed_block_timeout);
+
+                tokio::pin!(timer);
+
+                // Get the highest of all the results
+                let mut total_stake = 0;
+                let mut highest_round = 0;
+                'main_loop: loop {
+                    tokio::select! {
+                        Some((result, authority_index)) = results.next() => {
```

Comment:
I think it is less error prone to get `result` from `results.next()` next and break if `result.is_none()`, instead of only breaking after processing a valid result and `results` become empty. In the current logic, it seems possible to have the last return from results being malformatted / invalid block. Then the loop will have to wait until timeout.

Reply:
That's good point. Not the end of the world in waiting for the timeout but it's always better to favour efficiency